pkg/api/steve/catalog/plugin.go

```diff
@@ -0,0 +1,106 @@
+package catalog
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	neturl "net/url"
+	"strings"
+
+	"github.com/gorilla/mux"
+	"github.com/rancher/rancher/pkg/controllers/dashboard/plugin"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+func SetupUIPluginHandlers(router *mux.Router) {
+	router.HandleFunc("/v1/uiplugins", indexHandler)
+	router.HandleFunc("/v1/uiplugins/{name}/{version}/{rest:.*}", pluginHandler)
+}
+
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	authed := isAuthenticated(r)
+	if authed {
+		index, err := json.Marshal(&plugin.Index)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			logrus.Error(err)
+		}
+		w.Write(index)
+	} else {
+		index, err := json.Marshal(&plugin.AnonymousIndex)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			logrus.Error(err)
+		}
+		w.Write(index)
+	}
+}
+
+func pluginHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	logrus.Debugf("http request vars %s", vars)
+	authed := isAuthenticated(r)
+	entry, ok := plugin.Index.Entries[vars["name"]]
+	if (!ok || entry.Version != vars["version"]) || (!authed && !entry.NoAuth) {
```

Comment:
I didn't return any specific error in case someone tries to fetch an extension that they don't have permission to because that would give away that the extension is installed in the cluster. By returning 404 the attacker won't know that the extension (that may have a vulnerability) is installed in the cluster

Reply:
I was thinking about this and I thought if we first check for auth first then the attacker won't know what was installed or not ? like here https://github.com/rancher/rancher/blob/release/v2.9/pkg/api/steve/catalog/operation.go#L57

Reply:
We still need to check the plugin that is being requested before returning an unauthorized error since it can allow access by unauthorized users